### PR TITLE
fix(rocjitsu): DBT host selection on multi-GPU systems

### LINKED ISSUE #9212: Fix rocjitsu DBT host selection on multi-GPU systems
## Problem

Rocjitsu DBT integration tests fail before entering their test bodies on systems with multiple GPUs of the requested host ISA. Checked-in DBT configs use `host_gpu_id: 0` for automatic selection, but the launcher rejects the configuration when more than one physical GPU matches `host_isa`.

On an eight-GPU gfx942 system, a representative failure is:

```text
rocjitsu: dbt_guest.host_isa 'gfx942' matches 8 host GPUs; set host_gpu_id to one of: ...
```

The same physical ambiguity check is also applied to simulator-backed DBT configurations, even though those configurations execute against simulated topology rather than a physical host GPU.

## Expected behavior

- `host_gpu_id: 0` selects the first matching GPU in the effective ROCm-visible order.
- Automatic selection honors `ROCR_VISIBLE_DEVICES`, then `HIP_VISIBLE_DEVICES` or its `CUDA_VISIBLE_DEVICES` fallback, including numeric ordering and supported UUID forms.
- One synthetic DBT guest maps consistently to one selected host agent; it must not independently choose a host in the launcher, KFD interposer, and HSA hook.
- Explicit nonzero `host_gpu_id` values remain authoritative and are rejected clearly when hidden or ISA-incompatible.
- Hardware and simulator backends resolve against their own topology.
- No private application-facing environment variable is introduced for the resolved GPU.

## Acceptance criteria

- Simulator- and hardware-backed DBT tests pass on homogeneous multi-GPU hosts.
- Standard ROCm visibility selectors preserve their documented ordering and precedence.
- The resolved KFD GPU ID is handed consistently to all DBT runtime layers.
- Unit coverage includes selector edge cases, handoff parsing, explicit selection failures, and automatic selection.

### PR DESCRIPTION
**Pull request by amd-arosa**
## Motivation

Rocjitsu DBT tests fail before entering their test bodies when multiple GPUs share the configured host ISA. Automatic `host_gpu_id: 0` selection was rejected as ambiguous, and simulator-backed DBT was incorrectly checked against physical KFD topology.

This change makes one synthetic DBT guest select one host consistently while respecting standard ROCm device visibility.

## Technical Details

- Resolve automatic host selection from the backend topology after applying `ROCR_VISIBLE_DEVICES`, then `HIP_VISIBLE_DEVICES` or its `CUDA_VISIBLE_DEVICES` fallback.
- Preserve ROCr numeric/UUID and CLR client-selector semantics in a focused, testable visibility helper.
- Validate explicit GPU pins for visibility and ISA compatibility.
- Atomically publish the config path and resolved KFD GPU ID through the existing per-invocation runtime handoff so the launcher, KFD interposer, and HSA hook share one selection.
- Support both physical and simulator execution topologies without a private application-facing environment variable.
- Add unit and integration coverage for selector ordering, UUIDs, malformed handoffs, explicit pins, simulator topology, and multi-GPU execution.

## Issue Tracking

Fixes #9212

## Test Plan

- Build `rocjitsu_bin`, `rocjitsu_shared`, `rocjitsu_hooks`, `rocjitsu_tests`, and `launch_preload_test`.
- Run all `ConfigLoaderTest`, `LaunchPreloadTest`, and `RocmVisibilityTest` tests.
- Run all affected simulator- and hardware-backed DBT integration tests.
- Exercise combined `ROCR_VISIBLE_DEVICES` and `HIP_VISIBLE_DEVICES` ordering on an eight-GPU gfx942 system.

## Test Result

- 52/52 config, launcher, and visibility tests passed; two sanitizer-runtime-specific tests were skipped as expected.
- 25/25 affected DBT integration tests passed on an eight-GPU gfx942 host.
- Representative simulator and combined ROCR/HIP selector probes passed.
- Pre-commit whitespace, EOF, clang-format, and gersemi hooks passed.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/rocm-systems/blob/develop/CONTRIBUTING.md.

### CODE REVIEW

**Comment by amd-arosa** [comment]
fixing merge conflicts

**Review by atgutier — COMMENTED** [review]
A couple agent items:

## MAJOR

  1. tools/rocjitsu/rocm_visibility.cpp:61-69,111-119 — malformed selectors are converted into valid device selections.

     The parser is more permissive than the ROCr and CLR implementations it models:
      - ROCR_VISIBLE_DEVICES=GPU- is invalid in ROCr. On a one-GPU topology, rocjitsu accepts it as a unique UUID prefix and rewrites it to 0,1, exposing both host and guest.
      - CLR requires numeric HIP/CUDA tokens to use canonical decimal spelling. HIP_VISIBLE_DEVICES=01 should terminate selection, but rocjitsu accepts ordinal 1 and normalizes it to 1.

     This is not merely diagnostic drift: invalid isolation settings that should expose no device can instead select and execute on a GPU.

     Fix: enforce ROCr’s UUID-token length requirements and, after parsing a client ordinal, require token == std::to_string(index). Add coverage for GPU-, 01, 00, and valid-prefix-then-malformed sequences.

     Confidence: high.

  2. lib/rocjitsu/src/rocjitsu/kmd/linux/interposer.cpp:168-181 — the interposer can consume a different GPU ID than the HSA hook.

     The launcher writes:

     <config path>
     <resolved GPU ID>

     The HSA-hook loader reads the complete file, but the interposer performs one read() capped at 4095 bytes and treats any positive result as complete. A valid near-PATH_MAX config path can place only part
     of the GPU ID inside that buffer. For example, the interposer can parse 2 while the HSA hook parses 28851; if the newline falls outside the buffer, the interposer instead performs its own automatic
     selection.

     That directly violates the PR’s central invariant that GuestKfd and the HSA hook share one resolved host. KFD operations may then be rewritten to one GPU while queues and translated code objects use
     another.

     Fix: read to EOF into a dynamic buffer, retry EINTR, fail on other read errors, and apply a deliberate maximum handoff size. Add a handoff exceeding 4095 bytes and verify both consumers obtain the
     identical ID.

     Confidence: high.

  ## MINOR

  1. tests/config_test.cpp:637-660 — the new runtime-handoff test uses the fixed shared directory /tmp/rocjitsu_runtime_config_handoff_test.

     Concurrent test processes can overwrite each other’s config_path or remove the directory while another process is reading it, producing flaky failures or cross-test configuration reads.

     Fix: use the existing test::ScopedTempDirectory helper and RAII cleanup.

     Confidence: high.

**Review by kuhar — COMMENTED** [review]
A few comments

File: emulation/rocjitsu/tests/config_test.cpp [hunk 1]
```diff
@@ -529,38 +571,142 @@ TEST(ConfigLoaderTest, LoadsSimulatorDbtBackendConfig) {
...
+  config::DbtGuestConfig dbt;
+  dbt.enabled = true;
+  EXPECT_THROW(
+      config::apply_resolved_dbt_host_gpu_id(dbt, *handoff->resolved_gpu_id, "test handoff"),
+      std::runtime_error);
+}
+
+TEST(ConfigLoaderTest, LoadsDbtRuntimeConfigHandoffFromInvocationDirectory) {
+  const std::filesystem::path runtime =
+      std::filesystem::temp_directory_path() / "rocjitsu_runtime_config_handoff_test";
```
Comment (kuhar) [inline emulation/rocjitsu/tests/config_test.cpp]: This is a cross-process test race. Both the runtime directory and config filename are fixed under /tmp, and 32 concurrent runs exited 123 with gpu_id 0, empty JSON, and remove_all failures. We should give each invocation a unique temporary directory, for example with getpid() or mkdtemp, and keep all fixture files inside it.

File: emulation/rocjitsu/tests/launch_preload_test.cpp [hunk 2]
```diff
@@ -98,6 +104,129 @@ TEST(LaunchPreloadTest, EnvpCacheRebuildsAfterSet) {
...
+  const auto expanded = rocjitsu::cli::expanded_rocr_visible_devices(test_gpus(), "GPU-2222");
+  ASSERT_TRUE(expanded);
+  EXPECT_EQ("1,3", *expanded);
+}
+
+TEST(RocmVisibilityTest, DoesNotExpandEmptyRocrSelection) {
+  EXPECT_FALSE(rocjitsu::cli::expanded_rocr_visible_devices(test_gpus(), "5"));
+}
+
+TEST(RocmVisibilityTest, SelectsFirstVisibleIsaMatch) {
```
Comment (kuhar) [inline emulation/rocjitsu/tests/launch_preload_test.cpp]: It would be useful to cover the advertised explicit-pin success path. The current tests cover automatic selection and all three error statuses, but not select_host_gpu(test_gpus(), 100, 90402) returning Selected with gpu_id 100. Please add a focused case so a regression in nonzero host_gpu_id routing is caught.

File: emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      }
+    }
+
+    uint32_t index = 0;
+    const char *begin = token.data();
+    const char *end = begin + token.size();
+    auto [ptr, error] = std::from_chars(begin, end, index);
+    if (error != std::errc{} || ptr != end || index >= gpus.size())
+      return filtered;
+    if (std::none_of(filtered.begin(), filtered.end(),
```
Comment (kuhar) [inline emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:163]: It would be useful to pin the intentionally different duplicate-token behavior in focused tests. The upstream ROCR parser terminates on a repeated ordinal, while CLR skips the duplicate and continues; these two filters match that behavior today, but neither duplicate path is covered. Please add 0,0,1 cases for both filters, and a client-side 2,0 case would also cover numeric reordering directly.

File: emulation/rocjitsu/tests/config_test.cpp [hunk 4]
```diff
@@ -529,38 +571,142 @@ TEST(ConfigLoaderTest, LoadsSimulatorDbtBackendConfig) {
...
+      write_temp_config("rocjitsu_runtime_dbt_config_test.json", R"({
+        "dbt_guest": {
+          "enabled": true,
+          "guest_isa": "gfx950",
+          "host_isa": "gfx942"
+        }
+      })");
+  {
+    std::ofstream handoff(runtime / "config_path");
+    handoff << config_path.string() << "\n28851\n";
```
Comment (kuhar) [inline emulation/rocjitsu/tests/config_test.cpp]: I think it would help to test the production handoff writer and parser as one contract. This fixture duplicates the two-line path and resolved-GPU format by hand, while write_config_file in tools/rocjitsu/main.cpp:175 owns the actual serializer and atomic rename. A round trip through the real writer would catch format drift; extracting the serialization beside parse_dbt_runtime_config_handoff would make that focused test possible.

File: emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                ? HostSelectionStatus::Selected
+                : HostSelectionStatus::ExplicitGpuIsaMismatch,
+            configured_gpu_id};
+  }
+
+  const auto match =
+      std::find_if(visible_gpus.begin(), visible_gpus.end(), [&](const VisibleGpu &gpu) {
+        return gpu.gfx_target_version == gfx_target_version;
+      });
+  return match == visible_gpus.end() ? HostSelection{HostSelectionStatus::NoIsaMatch, 0}
```
Comment (kuhar) [inline emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:251]: We should not leave a physical GPU ahead of the selected DBT host in the normalized client order. With [gfx1100 gpu_id=100, gfx942 gpu_id=101], HIP_VISIBLE_DEVICES=0,1, and host_isa=gfx942, select_host_gpu chooses 101 but normalized_client_visible_devices stays 0,1. The hook substitutes the guest only in gpu_id 101’s slot, so CLR still exposes gfx1100 as device 0 and an application using its default device bypasses DBT. Please either require the first client-visible GPU to match host_isa or rewrite the selector so the chosen host/guest occupies device 0, and add this heterogeneous-ISA case.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp [hunk 6]
```diff
@@ -176,11 +210,15 @@ std::optional<DbtGuestConfig> load_dbt_guest_config_from_runtime_config() {
...
-  std::getline(file, path);
-  if (path.empty())
+  const std::string contents((std::istreambuf_iterator<char>(file)),
+                             std::istreambuf_iterator<char>());
+  std::optional<DbtRuntimeConfigHandoff> handoff = parse_dbt_runtime_config_handoff(contents);
+  if (!handoff)
     return std::nullopt;
-  return load_dbt_guest_config_from_file(path);
+  DbtGuestConfig config = load_dbt_guest_config_from_file(handoff->config_path);
+  if (handoff->resolved_gpu_id)
```
Comment (kuhar) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp]: This would be safer if an automatic enabled DBT config failed closed when the handoff lacks a resolved GPU ID. A path-only handoff leaves host_gpu_id at 0; GuestKfd then picks the first raw-topology ISA match while the HSA hook picks the first ROCR-visible match. With identical GPUs and ROCR_VISIBLE_DEVICES=1,0, those can be different physical devices. Please require the second line when DBT is enabled with host_gpu_id == 0, while preserving path-only handoffs for non-DBT or explicitly pinned configs, and update the old fixtures plus add a rejection test.

**Review by kuhar — COMMENTED** [review]
A few comments

File: emulation/rocjitsu/tests/config_test.cpp (hunk 1 above)
Comment (kuhar) [reply emulation/rocjitsu/tests/config_test.cpp]: The merge makes the JSON config file unique, but it does not resolve this finding: the invocation directory is still the fixed `/tmp/rocjitsu_runtime_config_handoff_test` path. On the current head, 9 of 32 concurrent runs failed with either `loaded == false` or `host.gpu_id == 0`.

Please use the newly added `test::ScopedTempDirectory` for `runtime` and place `config_path` inside it. That follows the new `scoped_temp.h` convention and removes the manual `remove_all()` cleanup as well as the remaining cross-process race.

Comment (amd-arosa) [reply emulation/rocjitsu/tests/config_test.cpp]: 573e63db21f852772e4ea9eaa333ac7f9ca70761

File: emulation/rocjitsu/tests/launch_preload_test.cpp (hunk 2 above)
Comment (amd-arosa) [reply emulation/rocjitsu/tests/launch_preload_test.cpp]: 3c5853aa1cc880a332f7ae7e7740cd65cf880d76

File: emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp (hunk 3 above)
Comment (amd-arosa) [reply emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:163]: 2429986655b8f6afaa9a2f5abde8ed1cc88d47fd

File: emulation/rocjitsu/tests/config_test.cpp (hunk 4 above)
Comment (amd-arosa) [reply emulation/rocjitsu/tests/config_test.cpp]: a388c8c3aeb5dd282c635f28da3e3e561a72edb0

File: emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp (hunk 5 above)
Comment (amd-arosa) [reply emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:251]: 6e108c2db203b670e9b350284846805b39fe6c05

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp (hunk 6 above)
Comment (amd-arosa) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp]: 6f1255419c87f46768a889c80c9cc56f55f2e9f3

**Review by kuhar — COMMENTED** [review]
A few comments

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/hooks/rj_hsa_dbt_hooks.cpp [hunk 7]
```diff
@@ -1406,6 +1406,10 @@ RjHsaLayer &layer() {
 /// not expose that value directly, so the hook reads the redirected topology
 /// tree and later compares agents by `HSA_AMD_AGENT_INFO_DRIVER_NODE_ID`.
 [[nodiscard]] std::optional<uint32_t> node_id_for_kfd_gpu_id(uint32_t gpu_id) {
+  // Unit tests provide fake HSA node IDs and a matching synthetic topology tree.
+  if (const char *root = std::getenv("ROCJITSU_HSA_HOOK_TOPOLOGY_NODES_ROOT"); root && *root)
```
Comment (kuhar) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/hooks/rj_hsa_dbt_hooks.cpp]: This would be safer if the unit test used a test-only seam instead of adding ROCJITSU_HSA_HOOK_TOPOLOGY_NODES_ROOT to the production hook. When that variable is inherited or stale, node_id_for_kfd_gpu_id() consults only the supplied tree and never falls back to the real KFD roots, so it can reject a valid resolved host or map it to fabricated topology. Please keep the production root policy fixed and inject the synthetic resolver only for the unit test.

File: emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp (hunk 5 above)
Comment (kuhar) [reply emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:251]: We should make selected-host-at-zero independent of whether HIP_VISIBLE_DEVICES or CUDA_VISIBLE_DEVICES was already set. With gfx1100 gpu_id 100 followed by gfx942 gpu_id 101 and no client selector, automatic selection chooses 101, but normalized_client_visible_devices() returns nullopt, so rj_iterate_agents() leaves the physical gfx1100 at public device 0 and substitutes the guest only at device 1. Please synthesize an ordering override for this clientless heterogeneous-ISA case and add focused coverage.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp (hunk 6 above)
Comment (kuhar) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp]: The same fail-closed check still needs to cover the KFD interposer. load_dbt_guest_config_from_runtime_config() now rejects a path-only handoff for enabled automatic DBT, but InterposerContext::get_or_create() loads the config directly and can still construct GuestKfd with host.gpu_id == 0. Please share the handoff validation between both consumers, or mirror the guard before GuestKfd construction, and cover the interposer path-only case.

File: emulation/rocjitsu/tests/launch_preload_test.cpp (hunk 2 above)
Comment (kuhar) [reply emulation/rocjitsu/tests/launch_preload_test.cpp]: Please change the explicit-pin test to select gpu_id 101 rather than 100. Both the automatic path and the current explicit path return the first matching GPU, 100, so the test still passes if the nonzero pin is ignored. Pinning 101 makes the paths diverge: automatic selection returns 100, while a working explicit branch must return 101.

Comment (amd-arosa) [reply emulation/rocjitsu/tests/launch_preload_test.cpp]: e3d68beba40246c2edc2d754b3bb5d9564616999

File: emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp (hunk 5 above)
Comment (amd-arosa) [reply emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:251]: 4d974f71fea0d4513307fb4ea7da799b4403f0fc

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp (hunk 6 above)
Comment (amd-arosa) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp]: fe591a5cde

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/hooks/rj_hsa_dbt_hooks.cpp (hunk 7 above)
Comment (amd-arosa) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/hooks/rj_hsa_dbt_hooks.cpp]: 436245e989

Comment (amd-arosa) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/hooks/rj_hsa_dbt_hooks.cpp]: don't love this pattern. would be used to doing DI but unsure if that's clean to do in c++.

File: emulation/rocjitsu/tests/dbt/hsa_hooks_unit_test.cpp [hunk 8]
```diff
@@ -992,6 +1008,25 @@ TEST(HsaHooksUnitTest, IterateAgentsDropsGuestOwnSlotWhenHostAppearsFirst) {
...
+  std::vector<uint64_t> seen;
+  hsa_status_t status = api.core.hsa_iterate_agents_fn(
+      [](hsa_agent_t agent, void *data) -> hsa_status_t {
+        static_cast<std::vector<uint64_t> *>(data)->push_back(agent.handle);
+        return HSA_STATUS_SUCCESS;
+      },
+      &seen);
+
+  EXPECT_EQ(status, HSA_STATUS_SUCCESS);
+  EXPECT_EQ(seen, (std::vector<uint64_t>{kGuestAgent.handle, kHostAgent.handle}));
```
Comment (kuhar) [inline emulation/rocjitsu/tests/dbt/hsa_hooks_unit_test.cpp]: We should fail closed when the configured host GPU cannot be mapped to an HSA topology node. In this case `AgentMapper` records no mapping and `rj_iterate_agents()` calls the original iterator, exposing both physical agents; a default-device application can then bypass the DBT guest. Please make initialization or iteration fail and change this test to expect that failure.

**Review by kuhar — COMMENTED** [review]
We should mirror the malformed-UUID validation on the client path. In `emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:114-117`, `filter_client_visible_gpus` still treats any token containing `GPU-` as a substring match; against the current source, bare `GPU-` selects gpu_id 100 and `1,GPU-,2` exposes all three GPUs, while the ROCR path rejects it. Please require a valid `GPU-...` prefix and length, use prefix matching, and cover both inputs in `MalformedSelectorsTerminateSelection`.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/hooks/rj_hsa_dbt_hooks.cpp (hunk 7 above)
Comment (kuhar) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/hooks/rj_hsa_dbt_hooks.cpp]: This would be safer if the test seam stayed out of the shipped hook. In an isolated HEAD build, `nm -D` lists `rj_hsa_dbt_set_topology_nodes_root_for_test` beside `OnLoad` and `OnUnload`, so any in-process caller can redirect host identity lookup. Please isolate the seam in a test-only build or translation unit so the production hook exports only its HSA tool entry points.

File: emulation/rocjitsu/tests/launch_preload_test.cpp (hunk 2 above)
Comment (kuhar) [reply emulation/rocjitsu/tests/launch_preload_test.cpp]: We should keep this test pinned to gpu_id 101. The staged revert changes it to 100, which is also the automatic first match: the built selector returns 100 for both automatic selection and an explicit pin to 100. A missing or bypassed explicit-pin branch would therefore stay green; pinning 101 makes the two paths diverge and preserves the regression guard.

File: emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp (hunk 5 above)
Comment (kuhar) [reply emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:251]: The staged revert removes commit `4d974f71fe` and reopens this path. A direct probe against the on-disk tree returns `nullopt` for a clientless heterogeneous topology with selected gpu_id 101, while HEAD returns `HIP_VISIBLE_DEVICES=1,0`. Please keep the synthesized host-first override and its two focused tests.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp (hunk 6 above)
Comment (kuhar) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/config/dbt_guest_config.cpp]: The staged revert removes the shared handoff validator, so this applies again in the KFD interposer: a path-only enabled automatic handoff can construct `GuestKfd` with gpu_id 0 while the HSA consumer rejects the same bytes. It also removes the pinned-host ISA cross-check. Please keep `load_dbt_guest_config_from_handoff()`, the explicit-pin validation, and the focused rejection tests.

**Comment by amd-arosa** [comment]
> `emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:114-117`, `filter_client_visible_gpus` still treats any token containing `GPU-` as a substring match; against the current source, bare `GPU-` selects gpu_id 100 and `1,GPU-,2` exposes all three GPUs, while the ROCR path rejects it. Please require a valid `GPU-...` prefix and length, use prefix matching, and cover both inputs in `MalformedSelectorsTerminateSelection`.

are you sure?

`if (token.size() >= 5 && token.size() <= 20 && token.starts_with("GPU-") && token != "GPU-XX") {`

and the test 

[TEST(RocmVisibilityTest, MalformedSelectorsTerminateSelection) {
  EXPECT_TRUE(rocjitsu::cli::filter_rocr_visible_gpus(test_gpus(), "GPU-").empty());
  EXPECT_TRUE(rocjitsu::cli::filter_client_visible_gpus(test_gpus(), "01").empty());
  EXPECT_TRUE(rocjitsu::cli::filter_client_visible_gpus(test_gpus(), "00").empty());

  const auto rocr_prefix = rocjitsu::cli::filter_rocr_visible_gpus(test_gpus(), "0,GPU-,1");
  ASSERT_EQ(1u, rocr_prefix.size());
  EXPECT_EQ(100u, rocr_prefix[0].gpu_id);

  const auto client_prefix = rocjitsu::cli::filter_client_visible_gpus(test_gpus(), "1,00,2");
  ASSERT_EQ(1u, client_prefix.size());
  EXPECT_EQ(101u, client_prefix[0].gpu_id);
}](https://github.com/ROCm/rocm-systems/blob/436245e9893f36b544b4a572ea78b95a40be1f56/emulation/rocjitsu/tests/launch_preload_test.cpp#L163-L175)

File: emulation/rocjitsu/tests/dbt/hsa_hooks_unit_test.cpp (hunk 8 above)
Comment (amd-arosa) [reply emulation/rocjitsu/tests/dbt/hsa_hooks_unit_test.cpp]: I think this is stale?

File: emulation/rocjitsu/tests/guest_kfd_test.cpp [hunk 9]
```diff
@@ -151,12 +152,47 @@ bool install_inline_dbt_config(std::string simulator_json, const char *host_isa,
...
+  const auto json = read_active_config_json();
+  if (!dir || !json)
+    return false;
+
+  try {
+    std::filesystem::path config_dir = *dir;
+    constexpr size_t kTargetParentLength = 4081;
+    while (config_dir.string().size() + 201 <= kTargetParentLength)
+      config_dir /= std::string(200, 'a');
+    const size_t final_component_length = kTargetParentLength - config_dir.string().size() - 1;
```
Comment (kuhar) [inline emulation/rocjitsu/tests/guest_kfd_test.cpp]: I think this may end up being an unsigned underflow? If the starting directory length is 61 modulo 201, the loop stops at exactly 4081 bytes and 4081 - 4081 - 1 wraps to SIZE_MAX; std::string then throws length_error outside the filesystem_error catch. Please compute and bounds-check the remaining length before subtracting, and report which size invariant failed instead of returning a bare false.

**Review by kuhar — COMMENTED** [review]
We should apply the UUID spelling checks to filter_client_visible_gpus as well. The size and starts_with guard quoted in the existing discussion is only in the ROCR branch; emulation/rocjitsu/tools/rocjitsu/rocm_visibility.cpp:114-117 still uses substring matching, so bare GPU- selects gpu_id 100 and 1,GPU-,2 produces 101,100,102 at 7010b4e5. Please require a valid GPU- prefix and length there and add both cases to MalformedSelectorsTerminateSelection.

File: emulation/rocjitsu/tests/dbt/hsa_hooks_unit_test.cpp (hunk 8 above)
Comment (kuhar) [reply emulation/rocjitsu/tests/dbt/hsa_hooks_unit_test.cpp]: This still applies at 7010b4e5: MissingResolvedHostKeepsPhysicalAgentOrder expects successful iteration returning both the guest and physical host when node_id_for_kfd_gpu_id cannot map the configured host. GuestKfd now rejects an unmatched host with ENODEV, but the HSA hook still falls back to the original iterator. We should make HSA initialization or iteration fail too and update this test to expect that failure.

Comment (amd-arosa) [reply emulation/rocjitsu/tests/dbt/hsa_hooks_unit_test.cpp]: ddab036

File: emulation/rocjitsu/tests/guest_kfd_test.cpp (hunk 9 above)
Comment (amd-arosa) [reply emulation/rocjitsu/tests/guest_kfd_test.cpp]: f1270ba

**Review by kuhar — APPROVED** [review]
LGTM
